### ISSUE TITLE
property type wrong

### ISSUE THREAD

**Issue by ray007** [google/closure-compiler#2439]
I have several boolean properties defined with `Object.defineProperties` on my class prototype returning the value with a `get` function. Unfortunately, I cannot use them as parameter where a boolean is required:
```
myfile.js:309: ERROR - actual parameter 2 of Date.prototype.strftime does not match formal parameter
found   : {get: function (this:myCtrl): boolean}
required: (boolean|undefined)
	return this.format ? d.strftime(this.format, this.showUtc) : d.toString();
	                                             ^^^^^^^^^^^^
```
Casting to `boolean` also does not work, though coercing the value with `!!` does work.

**Comment by ChadKillingsworth** [comment]
You need to post exactly how you are defining these.

**Comment by ray007** [comment]
Defining properties (reduced to 1) is done as follows:
```
Object.defineProperties(myCtrl.prototype, /** @lends {myCtrl.prototype} */ {
	showUtc: {
		/** @this {myCtrl} */
		get: function() { return (this._ctrlFlags & CF_SHOWUTC) == CF_SHOWUTC; },
	},
});
```

**Comment by ChadKillingsworth** [comment]
Does this work?

```js
Object.defineProperties(myCtrl.prototype, /** @lends {myCtrl.prototype} */ {
	showUtc: {
		/**
                  * @return {boolean}
                  * @this {myCtrl}
                  */
		get: function() { return (this._ctrlFlags & CF_SHOWUTC) == CF_SHOWUTC; },
	},
});
```

**Comment by ray007** [comment]
No, I already tried with `@return`, only changes the return type in the error message.

**Comment by brad4d** [comment]
@dimvar what kind of annotation is needed to make sure the type checker correctly recognizes a case like this?

**Comment by dimvar** [comment]
I tried an [example](https://closure-compiler-debugger.appspot.com/#input0%3Dclass%2520Foo%2520%257B%250A%2520%2520constructor()%2520%257B%2520this.prop%2520%253D%2520123%253B%2520%257D%250A%2520%2520%252F**%2520%2540return%2520%257Bnumber%257D%2520*%252F%250A%2520%2520get%2520propGetter()%2520%257B%2520return%2520this.prop%253B%2520%257D%250A%257D%250A%250AObject.defineProperties(Foo.prototype%252C%250A%2520%2520%252F**%2520%2540lends%2520%257BFoo.prototype%257D%2520*%252F%2520%257B%250A%2520%2520%2520%2520%252F**%2520%2540type%2520%257Bnumber%257D%2520*%252F%250A%2520%2520%2520%2520num2%253A%2520123%252C%250A%2520%2520%2520%2520%252F**%2520%2540return%2520%257Bnumber%257D%2520*%252F%250A%2520%2520%2520%2520get%2520getter2()%2520%257B%2520return%2520this.prop%253B%2520%257D%250A%2520%2520%257D)%253B%250A%250Avar%2520%252F**%2520string%2520*%252F%2520s%253B%250A%250As%2520%253D%2520(new%2520Foo).propGetter%253B%250A%250As%2520%253D%2520(new%2520Foo).num2%253B%250A%250As%2520%253D%2520(new%2520Foo).getter2%253B%26input1%26conformanceConfig%26externs%26refasterjs-template%26includeDefaultExterns%3D1%26CHECK_SYMBOLS%3D1%26MISSING_PROPERTIES%3D1%26TRANSPILE%3D1%26CHECK_TYPES%3D1%26CLOSURE_PASS%3D1%26PRESERVE_TYPE_ANNOTATIONS%3D1%26PRETTY_PRINT%3D1) with lends and it works as expected. Can you check what your code does differently from my example?

**Comment by ray007** [comment]
Take a look at [this example](https://closure-compiler-debugger.appspot.com/#input0%3D%252F**%2520%2540constructor%2520*%252F%250Afunction%2520Foo()%2520%257B%250A%2520%2520%2520%2520this.prop%2520%253D%2520true%253B%250A%257D%253B%250A%250A%252F**%2520%2540param%2520%257Bboolean%257D%2520x%2520*%252F%250AFoo.prototype.test%2520%253D%2520function(x)%2520%257B%250A%2520%2520%2520%2520return%2520%27%27%2520%252B%2520!!x%253B%250A%257D%253B%250A%250AObject.defineProperties(Foo.prototype%252C%250A%2520%2520%252F**%2520%2540lends%2520%257BFoo.prototype%257D%2520*%252F%2520%257B%250A%2520%2520%2520%2520%252F**%2520%2540type%2520%257Bnumber%257D%2520*%252F%250A%2520%2520%2520%2520num2%253A%2520123%252C%250A%2520%2520%2520%2520getter2%253A%2520%257B%250A%2520%2520%2520%2520%2520%2520%252F**%2520%2540this%2520%257BFoo%257D%2520%2540returns%2520%257Bboolean%257D%2520*%252F%250A%2520%2520%2520%2520%2520%2520get%253A%2520function()%2520%257B%2520return%2520this.prop%253B%2520%257D%252C%250A%2520%2520%2520%2520%2520%2520%252F**%2520%2540this%2520%257BFoo%257D%2520*%252F%250A%2520%2520%2520%2520%2520%2520set%253A%2520function(x)%2520%257B%2520this.prop%2520%253D%2520x%253B%2520%257D%250A%2520%2520%2520%2520%257D%250A%2520%2520%257D)%253B%250A%250Avar%2520%252F**%2520string%2520*%252F%2520s%253B%250Avar%2520%252F**%2520Foo%2520*%252F%2520f%253B%250A%250Af%2520%253D%2520new%2520Foo()%253B%250As%2520%253D%2520f.test(f.getter2)%253B%250A%26input1%26conformanceConfig%26externs%26refasterjs-template%26includeDefaultExterns%3D1%26CHECK_SYMBOLS%3D1%26CHECK_SUSPICIOUS_CODE%3D1%26MISSING_PROPERTIES%3D1%26CHECK_GLOBAL_NAMES%3D1%26CHECK_DEPRECATED%3D1%26CHECK_VISIBILITY%3D1%26CHECK_CONSTANTS%3D1%26CHECK_ES5_STRICT%3D1%26CHECK_GLOBAL_THIS%3D1%26TRANSPILE%3D1%26CHECK_TYPES%3D1%26CLOSURE_PASS%3D1%26PRESERVE_TYPE_ANNOTATIONS%3D1%26PRETTY_PRINT%3D1).
The warning is an error in my case, since I set almost all warning levels on compilation to error.

**Comment by dimvar** [comment]
The way you wrote the getter and setter is not correct. If you write it like [this](https://closure-compiler-debugger.appspot.com/#input0%3D%252F**%2520%2540constructor%2520*%252F%250Afunction%2520Foo()%2520%257B%250A%2520%2520%2520%2520this.prop%2520%253D%2520true%253B%250A%257D%253B%250A%250A%252F**%2520%2540param%2520%257Bboolean%257D%2520x%2520*%252F%250AFoo.prototype.test%2520%253D%2520function(x)%2520%257B%250A%2520%2520%2520%2520return%2520''%2520%252B%2520!!x%253B%250A%257D%253B%250A%250AObject.defineProperties(Foo.prototype%252C%250A%2520%2520%252F**%2520%2540lends%2520%257BFoo.prototype%257D%2520*%252F%2520%257B%250A%2520%2520%2520%2520%252F**%2520%2540type%2520%257Bnumber%257D%2520*%252F%250A%2520%2520%2520%2520num2%253A%2520123%252C%250A%2520%2520%2520%2520%252F**%2520%2540this%2520%257BFoo%257D%2520%2540returns%2520%257Bboolean%257D%2520*%252F%250A%2520%2520%2520%2520get%2520getter2()%2520%257B%2520return%2520this.prop%253B%2520%257D%252C%250A%2520%2520%2520%2520%252F**%2520%2540this%2520%257BFoo%257D%2520*%252F%250A%2520%2520%2520%2520set%2520getter2(x)%2520%257B%2520this.prop%2520%253D%2520x%253B%2520%257D%250A%2520%2520%257D)%253B%250A%250Avar%2520%252F**%2520string%2520*%252F%2520s%253B%250Avar%2520%252F**%2520Foo%2520*%252F%2520f%253B%250A%250Af%2520%253D%2520new%2520Foo()%253B%250As%2520%253D%2520f.test(f.getter2)%253B%250A%26input1%26conformanceConfig%26externs%26refasterjs-template%26includeDefaultExterns%3D1%26CHECK_SYMBOLS%3D1%26CHECK_SUSPICIOUS_CODE%3D1%26MISSING_PROPERTIES%3D1%26CHECK_GLOBAL_NAMES%3D1%26CHECK_DEPRECATED%3D1%26CHECK_VISIBILITY%3D1%26CHECK_CONSTANTS%3D1%26CHECK_ES5_STRICT%3D1%26CHECK_GLOBAL_THIS%3D1%26TRANSPILE%3D1%26CHECK_TYPES%3D1%26CLOSURE_PASS%3D1%26PRESERVE_TYPE_ANNOTATIONS%3D1%26PRETTY_PRINT%3D1), then type checking works as expected.

**Comment by ray007** [comment]
It's nice this syntax works.
But it's not the official syntax, see the [MDN Docs](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Global_Objects/Object/defineProperties).
So if I want this to work, I need to use the shortcut too?

**Comment by dimvar** [comment]
You're right, I was writing the object literal incorrectly. The type checker needs to look for getters/setters differently when the object literal is an argument to `Object.defineProperties`.

**Comment by ray007** [comment]
While this did seem fixed for a while, now that I've converted my code to modules and issue #2639 was fixed, this is a problem again.

**Comment by brad4d** [comment]
Created Google internal bug b/119819950

**Comment by ray007** [comment]
Very strange, how did I unassign @dimvar from this issue?
Don't remember doing that, and I don't think I even could do that.

Is there a way for me to view that "internal google bug"?

**Comment by lazarljubenovic** [comment]
@ray007 GitHub updates the assignee when a comment is posted if the assignee for some reason doesn't have privileges to be assigned to the issue (maybe not the part of the GitHub org anymore or something). Technically you did the action because your comment triggered the "check", but, indeed, very confusing,

**Comment by lauraharker** [comment]
@ray007 Sorry, we don't currently have a way for you to view the internal Google bug. We're working on improving coordination between GitHub and our internal bug-tracking system.

**Comment by ray007** [comment]
I didn't see anything in the changelog addressing this issue, but I did try the latest release 20190121, and most problems seem to be gone now. But not all.

I have (among other property definitions):
```
Object.defineProperties(myControl.prototype, /** @lends myControl.prototype */ {
	// some more properties
	// inverted flag
	_enabled: {
		/** @this {myControl} */
		get: function() {
			return !(this._appFlags & appFlags.CMF_DISABLED);
		},
		/** @this {myControl} */
		set: function(value) {
			if (value)
				this._appFlags &= ~appFlags.CMF_DISABLED;
			else
				this._appFlags |= appFlags.CMF_DISABLED;
		}
	},
	// inverted flag
	_visible: {
		/** @this {myControl} */
		get: function() {
			return !(this._appFlags & appFlags.CMF_HIDDEN);
		},
		/** @this {myControl} */
		set: function(value) {
			if (value)
				this._appFlags &= ~appFlags.CMF_HIDDEN;
			else
				this._appFlags |= appFlags.CMF_HIDDEN;
		}
	},
	// some more properties
});
```

the `_enabled` property seems to work fine, `_visible` apparently not:
```
[java] www/js/controls.js:949: WARNING - inconsistent return type
[java] found   : (boolean|{get: function(this:module$controls.default.myControl): ?, set: function(this:module$controls.default.myControl, ?): undefined})
[java] required: boolean
[java]     return r;
[java]            ^
[java]
[java] www/js/controls.js:1711: ERROR - condition always evaluates to true
[java] left : {get: function(this:module$controls.default.myControl): ?, set: function(this:module$controls.default.myControl, ?): undefined}
[java] right: boolean
[java]     var chg = (this._visible !== vis);
[java]                ^^^^^^^^^^^^^^^^^^^^^
```

The method for the warning in line 949 is:
```
/**
 * does this control want to process/consume a touchevent?
 * @param {Event} ev
 * @param {?{x:number, y:number}} p
 * @returns {boolean}
 */
myControl.prototype.wantsTouchEv = function(ev, p) {
	var r = this._visible && this._enabled;
	if (r && p)
		r = this.useTouch(ev, p);
	return r;
};
```

Not sure what's going wrong here, consider me confused.

**Comment by ray007** [comment]
Interestingly enough, if I change the check from `!==` to `!=`, the error in line 1711 goes away.

The warning in line 949 can be avoided by adding a cast to the functions first line:
```
var r = /** @type {boolean} */ (this._visible && this._enabled);
```

To summarize: I can make the warnings go away, but there's still something not quite right with the types.
